client/asset/eth/eth.go

```diff
@@ -670,12 +691,155 @@ func (eth *ExchangeWallet) unlockFunds(coins asset.Coins) error {
+}
+
+// SignedRefund returns an empty byte array. ETH does not support a pre-signed
+// redeem script becuase the nonce needed in the transaction can not be previously
+// determined. We will need to come up with a different way to support manual
+// refunds in ETH.
+func (*swapReceipt) SignedRefund() dex.Bytes {
+	return dex.Bytes{}
+}
+
+var _ asset.Receipt = (*swapReceipt)(nil)
+
+// Swap sends the swaps in a single transaction. The Receipts returned can
+// be used to refund a failed transaction. The fees used returned are the max
+// fees that will possibly be used, since in ethereum with EIP-1559 we cannot
+// know exactly how much fees will be used.
+func (eth *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint64, error) {
+	initiations := make([]dexeth.ETHSwapInitiation, 0, len(swaps.Contracts))
+	receipts := make([]asset.Receipt, 0, len(swaps.Contracts))
+
+	eth.lockedFundsMtx.Lock()
+	defer eth.lockedFundsMtx.Unlock()
+
+	var totalInputValue uint64
+	for _, input := range swaps.Inputs {
+		if _, ok := eth.lockedFunds[input.ID().String()]; !ok {
+			return nil, nil, 0,
+				fmt.Errorf("Swap: attempting to use coin that was not locked")
+		}
+		totalInputValue += input.Value()
+	}
+
+	var totalContractValue uint64
+	secretHashMap := make(map[string]bool)
+	for _, contract := range swaps.Contracts {
+		totalContractValue += contract.Value
+
+		if len(contract.SecretHash) != 32 {
+			return nil, nil, 0,
+				fmt.Errorf("Swap: expected secret hash of length 32, but got %v", len(contract.SecretHash))
+		}
+		encSecretHash := hex.EncodeToString(contract.SecretHash)
+		if _, ok := secretHashMap[encSecretHash]; ok {
+			return nil, nil, 0, fmt.Errorf("Swap: secret hashes must be unique")
+		}
+		secretHashMap[encSecretHash] = true
+		var secretHash [32]byte
+		copy(secretHash[:], contract.SecretHash)
+
+		if !common.IsHexAddress(contract.Address) {
+			return nil, nil, 0, fmt.Errorf("Swap: address in contract is not valid %v", contract.Address)
+		}
+		address := common.HexToAddress(contract.Address)
+
+		initiations = append(initiations,
+			dexeth.ETHSwapInitiation{
+				RefundTimestamp: big.NewInt(0).SetUint64(contract.LockTime),
+				SecretHash:      secretHash,
+				Participant:     address,
+				Value:           big.NewInt(0).SetUint64(contract.Value),
+			})
+	}
+
+	initGas, err := eth.getInitGas(len(initiations))
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	totalUsedValue := totalContractValue + (initGas * swaps.FeeRate)
+	if totalInputValue < totalUsedValue {
+		return nil, nil, 0, fmt.Errorf("Swap: coin inputs value %d < required %d", totalInputValue, totalUsedValue)
+	}
+
+	opts := bind.TransactOpts{
+		From:      eth.acct.Address,
+		Value:     big.NewInt(0).SetUint64(totalContractValue),
+		GasFeeCap: big.NewInt(0).SetUint64(swaps.FeeRate),
+		GasTipCap: big.NewInt(GasTipCap),
+		Context:   eth.ctx}
+
+	tx, err := eth.node.initiate(&opts, eth.networkID, initiations)
```

Comment:
Do we have a bigger problem with EIP-1559.  If we lock value (the "change" amount) based on `getInitGas` and `swaps.FeeRate` (the `GasFeeCap`), then `tx` gets mined using some unknown lower gas cost, the actual amount available is something else.
I suppose this just means we'll have more locked than is necessary until the order is fully executed and the change is returned?

Reply:
That's right. We could just use a legacy transaction and all the funds could be used up and this issue wouldn't exist, but I think we might as well let the user save a little bit.

Reply:
OK, we'll use an eip 1559 txn, but let's doc the unknown actual cost issue somewhere in this function.